samples/nrf9160/azure_fota/src/main.c

```diff
@@ -0,0 +1,248 @@
+		k_sem_give(&network_connected_sem);
+		network_connected = true;
+		break;
+	case LTE_LC_EVT_PSM_UPDATE:
+		printk("PSM parameter update: TAU: %d, Active time: %d\n",
+		       evt->psm_cfg.tau, evt->psm_cfg.active_time);
+		break;
+	case LTE_LC_EVT_EDRX_UPDATE: {
+		char log_buf[60];
+		ssize_t len;
+
+		len = snprintf(log_buf, sizeof(log_buf),
+			       "eDRX parameter update: eDRX: %f, PTW: %f",
+			       evt->edrx_cfg.edrx, evt->edrx_cfg.ptw);
+		if (len > 0) {
+			printk("%s\n", log_buf);
+		}
+		break;
+	}
+	case LTE_LC_EVT_RRC_UPDATE:
+		printk("RRC mode: %s\n",
+			evt->rrc_mode == LTE_LC_RRC_MODE_CONNECTED ?
+			"Connected" : "Idle");
+		break;
+	case LTE_LC_EVT_CELL_UPDATE:
+		printk("LTE cell changed: Cell ID: %d, Tracking area: %d\n",
+			evt->cell.id, evt->cell.tac);
+		break;
+	default:
+		break;
+	}
+}
+
+static void modem_configure(void)
+{
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already configured and LTE connected. */
+	} else {
+		int err = lte_lc_init_and_connect_async(lte_handler);
+
+		if (err) {
+			printk("Modem could not be configured, error: %d\n",
+			       err);
+			return;
+		}
+	}
+}
+
+/**@brief Configures AT Command interface. */
+static void at_configure(void)
+{
+	int err;
+
+	err = at_notif_init();
+	__ASSERT(err == 0, "AT Notify could not be initialized.");
+	err = at_cmd_init();
+	__ASSERT(err == 0, "AT CMD could not be established.");
+}
+
+void main(void)
+{
+	int err;
+
+	printk("Azure FOTA sample started\n");
+	printk("Initializing BSDlib\n");
+	printk("This may take a while if a modem firmware update is pending\n");
+
+	err = bsdlib_init();
+	switch (err) {
+	case MODEM_DFU_RESULT_OK:
+		printk("Modem firmware update successful!\n");
+		printk("Modem will run the new firmware after reboot\n");
+		k_thread_suspend(k_current_get());
+		break;
+	case MODEM_DFU_RESULT_UUID_ERROR:
+	case MODEM_DFU_RESULT_AUTH_ERROR:
+		printk("Modem firmware update failed\n");
+		printk("Modem will run non-updated firmware on reboot.\n");
+		break;
+	case MODEM_DFU_RESULT_HARDWARE_ERROR:
+	case MODEM_DFU_RESULT_INTERNAL_ERROR:
+		printk("Modem firmware update failed\n");
+		printk("Fatal error.\n");
+		break;
+	case -1:
+		printk("Could not initialize bsdlib.\n");
+		printk("Fatal error.\n");
+		return;
+	default:
+		break;
+	}
+
+	printk("BSDlib initialized\n");
+
+	k_delayed_work_init(&reboot_work, reboot_work_fn);
+	at_configure();
+
+	err = azure_iot_hub_init(NULL, azure_event_handler);
+	if (err) {
+		printk("Azure IoT Hub could not be initialized, error: %d\n",
+		       err);
+		return;
+	}
+
+	printk("Connecting to LTE network\n");
+	modem_configure();
+	k_sem_take(&network_connected_sem, K_FOREVER);
+
+	err = azure_iot_hub_connect();
+	if (err < 0) {
+		printk("azure_iot_hub_connect failed: %d\n", err);
+		return;
+	}
+
+	/* All initializations were successful mark image as working so that we
+	 * will not revert upon reboot.
+	 */
+	boot_write_img_confirmed();
```

Comment:
I'll leave it here so that it's easy to spot for users of the sample, but it now waits on a semaphore that's given in the `AZURE_IOT_HUB_READY_EVENT` case.